AnkiDroid/src/main/java/com/ichi2/anki/instantnoteeditor/InstantEditorViewModel.kt

```diff
@@ -0,0 +1,174 @@
+        withCol {
+            _deckId.postValue(config.get(Decks.CURRENT_DECK) ?: 1L)
+        }
+    }
+
+    suspend fun changeToClozeType() {
+        Timber.d("Changing to cloze type note")
+        withCol {
+            val model = notetypes.all().first { noteType ->
+                noteType.isCloze
+            }
+
+            _currentlySelectedNotetype.postValue(model)
+
+            val oldModelId = notetypes.current().getLong("id")
+            Timber.i("Changing note type to '%d", model.id)
+            viewModelScope.launch {
+                oldNoteId.emit(oldModelId)
+            }
+            setModel(this, model.id)
+        }
+    }
+
+    suspend fun getCol(): Collection {
+        return withCol { this }
+    }
+
+    fun checkAndSaveNote(activity: Activity) {
+        viewModelScope.launch {
+            val noteFieldsCheck = editorNote.value?.let { checkNoteFieldsResponse(it) }
+            if (noteFieldsCheck is NoteFieldsCheckResult.Failure) {
+                // TODO: check if we need more errors here like note editor
+                instantEditorError.emit(noteFieldsCheck.getLocalizedMessage(activity))
+                setSavingNote(false)
+                return@launch
+            }
+            Timber.d("Note fields check successful, saving note")
+            instantEditorError.emit(null)
+            saveWithProgress(activity)
+        }
+    }
+
+    suspend fun saveWithProgress(activity: Activity) {
+        putDeckIdInNote()
+        activity.withProgress(activity.resources.getString(R.string.saving_facts)) {
+            val note = editorNote.value ?: return@withProgress
+            val deckId = deckId.value ?: return@withProgress
+            undoableOp { addNote(note, deckId) }
+        }
+        noteSaved.emit(true)
+    }
+
+    fun setModel(col: Collection, modelId: Long) {
+        val model = col.notetypes.get(modelId)
+        if (model == null) {
+            Timber.w("New model %s not found, not changing note type", modelId)
+            return
+        }
+
+        col.notetypes.setCurrent(model)
+        val currentDeck = col.decks.current()
+        currentDeck.put("mid", modelId)
+        col.decks.save(currentDeck)
+
+        // refresh the note
+        _editorNote.postValue(
+            col.run {
+                val notetype = notetypes.current()
+                Note.fromNotetypeId(notetype.id)
+            }
+        )
+    }
+
+    private fun putDeckIdInNote() {
+        _editorNote.value?.notetype?.put("did", deckId)
+    }
+
+    fun getEditorNote(): Note? {
+        return editorNote.value
+    }
+
+    suspend fun checkClozeNoteTypes() {
+        viewModelScope.launch {
+            clozeNoteType.emit(
+                withCol {
+                    notetypes.all().any {
+                        Timber.d("Cloze note type: ${it.isCloze}")
+                        it.isCloze
+                    }
+                }
+            )
+        }
+    }
```

Comment:
+ this doesn't need to be suspect if using `viewModelScope.launch`

```suggestion
   fun checkClozeNoteTypes() = viewModelScope.launch {
        val hasClozeNoteType = withCol { notetypes.all().any { it.isCloze } }
        clozeNoteType.emit(hasClozeNoteType)
    }
```